### ISSUE TITLE
Portins() fails to change RequestedFocDate for FOC'd Automated Ports

### ISSUE THREAD

**Issue by ooglek** [Bandwidth/php-bandwidth-iris#73]
I tried to update the FOC date to a future date using the SDK, but it failed with an error:

```
Unable to modify actual FOC date for AUTOMATED port type
```

Doing so using the API directly, and not the PHP SDK, succeeded.

May be related to https://github.com/Bandwidth/php-bandwidth-iris/issues/23 and the change in July 2019 to add ActualFocDate to the Portin class model.

The API directly allows you to change the `RequestedFocDate` of an Automated Port that is in Status `FOC`. I tested this in BW Request #4013154 with AJ and David.

If you submit the XML for an Automated Port in FOC status:
```
<LnpOrderSupp>
 <RequestedFocDate>2021-08-16T17:30:00.000Z</RequestedFocDate>
</LnpOrderSupp>
```

it succeeds.

However, in the SDK, it fails, and `ActualFocDate` seems to be why.

AJ did a test on an internal order:

```
<LnpOrderSupp>
 <ActualFocDate>2021-08-16T17:30:00.000Z</ActualFocDate>
</LnpOrderSupp>
```

Response:
```
 <Errors>
 <Code>7368</Code>
 <Description>Unable to modify actual FOC date for INTERNAL port type</Description>
 </Errors>
```

This seems likely the culprit, as the code takes the whole `portin` order model as an array on `update()` and passes it to the server, which in the case of a port in FOC status, would include `ActualFocDate`, and thus get the error above.

https://github.com/Bandwidth/php-bandwidth-iris/blob/46c4b5fcbc59a1bb4f4c91aadf11ee2e7f57dcf0/src/PortinsModel.php#L124-L127

My process was roughly this:
```
$bw = new BandwidthAPIext(); // custom auth wrapper
$portin = $bw->account->portins()->portin($orderid);
$portData["RequestedFocDate"] = gmdate('c', $newDate);
$portin->set_data($portData);
$portin->update();
```

IMO there are two straight-forward solutions:

1. Modify the server to ignore ActualFocDate in PUT (`update()`) calls, because really, this is a read-only value anyway. Or heck, for sanity sake, only ignore it if ActualFocDate is the same on the BW side as the API request side, throw an error otherwise, since the API requester data may be out of date at the time of the PUT.

2. Modify the SDK to only submit changed values. This would require the SDK to hold the server-generated state of the `portin` class model, as well as the SDK modified model, do a diff, and then send the XML only with the changed values. While a simpler path would be to GET the portin prior to PUT and do a diff then, you've doubled the API traffic, and in production, the GET is probably done only a few seconds prior to the PUT anyway.

I'm sure there are other solutions, those are the two I came up with. Option 1 seems fastest and requires no changes on the API SDK user's end to fix. Option 2 might follow more closely the intent of the PUT API, to only pass fields you want to change, not the entire model. Or maybe Option 3 is something y'all come up with!

### Workaround

1. Make raw API calls directly using Guzzle/Curl. Limit PUT XML to only fields you want to change.
2. Use the SDK Raw Calls

I haven't actually done it in the SDK, but I think a combo that would work is:

```
$portin = $bw->account->portins()->portin($orderid);
$client = $portin->get_rest_client();
$client->put($portin->get_relative_namespace(), "LnpOrderSupp", ['RequestedFocDate' => '2021-06-07T18:00:00Z']);
```

**Comment by ajrice6713** [comment]
Hey Peter,

Thank you for the detailed issue here in addition to your work in the support ticket - we are about to begin a new sprint cycle and will take a look at this issue in the hopes of publishing a solution in this release cycle.

**Comment by ooglek** [comment]
Sounds good, thanks @ajrice6713 ! How long are your sprints, 2-weeks? Longer?

As I mentioned above:
- Fast Solution: Modify server to accept/ignore `ActualFocDate` in PUT requests so that legit changes to `RequestedFocDate` are allowed when a port is in `FOC` status.

- Slower and more complicated Solution: Modify SDK to only send fields changed from the last server response that populated the `Portin` class object being `update()`'d. 

Since the server-side is fast and I can't help, and the other solution may overcomplicate the SDK, I won't attempt a Pull Request. 😄

**Comment by ajrice6713** [comment]
We operate on 2 week sprints.

I actually suspect that the client side update would be quicker - the team that manages the server side resources for porting has a backlog of work much bigger than my team, so it may be easiest for us to just remove `<ActualFocDate>` from the payload on the PUT request.

One of my colleagues is already investigating the fix this afternoon.

**Comment by jmulford-bw** [comment]
Hey @ooglek can you share the code you're running that's causing this issue? I _think_ I know what's going on but just want to validate with what you're doing

**Comment by ooglek** [comment]
See the pseudo-code block in the original issue report @jmulford-bw about half way down, just after the paste of the `update()`

1. Pull the order
2. Update locally
3. call `->update()`

**Comment by jmulford-bw** [comment]
Hey @ooglek so the only problematic field was the `ActualFocDate` field, so we can get around this for now by just temporarily clearing it when doing the update request. https://github.com/Bandwidth/php-bandwidth-iris/pull/74 should fix this. We can definitely look into a server-client diff to make this a bit more robust at some point

**Comment by ooglek** [comment]
Excellent! Thanks @jmulford-bw -- sorry for my delayed appreciation. I see it has been released as Version 3.3.1 and I've got it installed. Huzzah!

I'll close this out.
